cpp/unreal_plugins/UrdfBot/Source/UrdfParser.h

```diff
@@ -0,0 +1,170 @@
+
+struct UrdfJointDesc;
+struct UrdfLinkDesc;
+struct UrdfMaterialDesc;
+
+enum class UrdfGeometryType
+{
+    Invalid = 0,
+    Box = 1,
+    Sphere = 2,
+    Cylinder = 3,
+    Mesh = 4,
+};
+
+enum class UrdfJointType
+{
+    Invalid = 0,
+    Revolute = 1,
+    Continuous = 2,
+    Prismatic = 3,
+    Fixed = 4,
+    Floating = 5,
+    Planar = 6
+};
+
+enum class CalibrationType
+{
+    Invalid = 0,
+    Rising = 1,
+    Falling = 2
+};
+
+struct UrdfGeometryDesc
+{
+    UrdfGeometryType type_;
+    FVector size_ = FVector(0.07); // required by Box, in [m]
+    float radius_ = 0.0352; // required by Sphere, Cylinder, in [m]
+    float length_ = 0.0206; // required by Cylinder, in [m]
+    std::string filename_; // required by Mesh
+    float scale_; // optional for Mesh
+};
+
+struct UrdfMaterialDesc
+{
+    std::string name_;
+
+    FVector4 color_;
+    std::string texture_;
+};
+
+struct UrdfInertialDesc
+{
+    FTransform origin_ = FTransform::Identity;
+    float mass_ = 0.0f;
+    Eigen::Matrix3f inertia_;
+};
+
+struct UrdfVisualDesc
+{
+    std::string name_;
+
+    FTransform origin_ = FTransform::Identity;
+    UrdfGeometryDesc geometry_;
+    UrdfMaterialDesc material_desc_;
+};
+
+struct UrdfCollisionDesc
+{
+    std::string name_;
+
+    FTransform origin_ = FTransform::Identity;
+    UrdfGeometryDesc geometry_;
+};
+
+struct UrdfLinkDesc
+{
+    std::string name_;
+
+    UrdfInertialDesc inertial_desc_;
+    std::vector<UrdfVisualDesc> visual_descs_;
+    std::vector<UrdfCollisionDesc> collision_descs_;
+
+    bool is_root_ = true;
+    std::vector<UrdfLinkDesc*> child_link_descs_ = {};
+    std::vector<UrdfJointDesc*> child_joint_descs_ = {};
+};
+
+struct UrdfJointDesc
+{
+    std::string name_;
+    UrdfJointType type_;
+
+    // transform from the parent link to the child link
+    FTransform origin_ = FTransform::Identity;
+    // parent link name
+    std::string parent_;
+    // child link name
+    std::string child_;
+    // joint axis specified in the joint frame
+    FVector axis_ = FVector(1, 0, 0);
+    // dynamics
+    float damping_ = 0.0f;
+    float friction_ = 0.0f;
+    // calibration
+    CalibrationType calibration_type_ = CalibrationType::Invalid;
+    float rising_ = 0.0f;
+    float falling_ = 0.0f;
+    // limits
+    float lower_ = 0.0f;
+    float upper_ = 0.0f;
+    float effort_ = 0.0f;
+    float velocity_ = 0.0f;
+    // mimic
+    std::string joint_;
+    float multiplier_ = 1.0f;
+    float offset_ = 0.0f;
+    // safety_controller
+    float soft_lower_limit_ = 0.0f;
+    float soft_upper_limit_ = 0.0f;
+    float k_position_ = 0.0f;
+    float k_velocity_ = 0.0f;
+};
+
+struct UrdfRobotDesc
+{
+    std::string name_;
+
+    std::map<std::string, UrdfLinkDesc> link_descs_ = {};
+    std::map<std::string, UrdfJointDesc> joint_descs_ = {};
+
+    UrdfLinkDesc* root_link_desc_ = nullptr;
+};
+
+class UrdfParser
+{
+public:
+    static UrdfRobotDesc parse(const std::string& file_name);
+
+private:
```

Comment:
```
private:
    // parse urdf nodes
    static UrdfGeometryDesc parseGeometryNode(FXmlNode* geometry_node);
    static UrdfMaterialDesc parseMaterialNode(FXmlNode* material_node);
    static UrdfInertialDesc parseInertialNode(FXmlNode* inertial_node);
    static UrdfVisualDesc parseVisualNode(FXmlNode* visual_node);
    static UrdfCollisionDesc parseCollisionNode(FXmlNode* collision_node);
    static UrdfLinkDesc parseLinkNode(FXmlNode* link_node);
    static UrdfJointDesc parseJointNode(FXmlNode* joint_node);
    static UrdfRobotDesc parseRobotNode(FXmlNode* robot_node);

    // parse basic parameters
    static FVector parseVector(const FString& string);
    static FVector4 parseVector4(const FString& string);
    static FQuat parseRotation(const FString& string);

    // parse enum types (seems like there need to be more of these enum functions)
    static UrdfJointType parseJointType(const FString& type);
};
```

Reply:
fixed

Reply:
`parseInertiaMatrixNode` seems removed here. Is it intented?

Reply:
Yes, I am suggesting inlining it into `parseInertialNode` because it doesn't match the rest of the functions here, and it is only ever called in that one location.

Reply:
Let's also inline `parseJointType`.

Reply:
fixed